### DISCUSSION TITLE
babel-plugin-transform-modules-commonjs and 'default'

### DISCUSSION THREAD

**Discussion by thernstig** [babel/babel#13358] [Q&A]
```
  System:
    OS: Linux 5.4 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz
  Binaries:
    Node: 14.16.0 - ~/.local/share/nvm/v14.16.0/bin/node
    npm: 6.14.11 - ~/.local/share/nvm/v14.16.0/bin/npm
  fs-extra: 8.1.0
```

A npm package `fs-extra` contains this code:

```js
const fs = require('graceful-fs')
...
...
...
if (typeof fs.realpath.native === 'function') {
  exports.realpath.native = u(fs.realpath.native)
}
```

I am using Jest to to mock the `fs` module, which worked fine before I recently upgrade to use ECMAScript modules. But I transpile the code with babel to change `import` statements into `require` with the babel config (which is recommended till Jest supports full ESM):

```json
{
  "plugins": [
    "@babel/plugin-transform-modules-commonjs"
  ]
}
```

I mock the `fs` module like:

```js
const fs = jest.createMockFromModule('fs');

fs.existsSync = jest.fn(() => true);

export default fs;
```

When running a test case I get:

```sh
TypeError: Cannot read property 'native' of undefined

      at Object.<anonymous> (node_modules/fs-extra/lib/fs/index.js:108:24)
      at Object.<anonymous> (node_modules/fs-extra/lib/index.js:6:3)
```

If I do a `console.log` of the `fs` object inside `server/node_modules/fs-extra/lib/fs/index.js` I can see it is an object like (note that I run this code without babel transpilation:

```js
{
  ...
  realpath: [Function: realpath] { native: [Function (anonymous)] }
  ...
}
```

But if I log it while running with Jest where it is transpiled with babel, I get:

```js
{
  default: {
    ...
    realpath: [Function: realpath] {
      _isMockFunction: true,
      ...
    }
    ...
  }
}
```

The problem here is then that `fs` contains a `default` property in the object, the module `fs-extra` fails since it cannot find `fs.realpath`.

I have no idea on how to solve this.

note: I am aware the latest version does not contain this, but the lib I am using will take a long time to update their dependencies

**Comment by nicolo-ribaudo** [comment]
I don't fully understand your setup, but I would try two things:
1. Use `import * as fs from "fs-extra"` instead of `import fs from "fs-extra"`
2. Or set the [`importInterop`](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#importinterop) option to `none`

If neither of those work, could you create a repository or CodeSandbox that shows the problem so that I can debug it?

**Reply by thernstig** [reply]
I am trying to create a repro, but then bumped it failing on another error. However, I reported the same issue at `node-fs-extra` and got a reply (see https://github.com/jprichardson/node-fs-extra/issues/905#issuecomment-847111313) to do a `module.exports = fs` instead of `export default fs` in my fs mock. And it worked. I have no idea why.

To answer your questions:
1. I am not doing any `import fs from "fs-extra"` myself. The only thing I am importing is `import fs from "fs"`. `fs-extra` is used by other node_module only.
2. I tried all variants of `importInterop` and non worked, the other two than default just spit out even more errors.

I will still try to create a repro tomorrow as I cannot share the current project, but still I find the finding above about using `module.exports = fs` very interesting.

**Reply by thernstig** [reply]
I _think_ I might know why this happens. `node_modules/fs-extra/lib/fs/index.js` is importing `const fs = require('graceful-fs')`, which imports `fs` in turn which I have mocked as a ECMAScript module.

`node_modules/fs-extra/lib/fs/index.js` has this piece of code:

```js
const fs = require('graceful-fs')
...
...
...
if (typeof fs.realpath.native === 'function') {
  exports.realpath.native = u(fs.realpath.native)
}
```

This means when it tries `fs.realpath.native` and since I mocked `fs`, it is an ECMASscript module. So the `@babel/plugin-transform-modules-commonjs` has exported all of fs's content on `fs.default.realpath.native` instead of `fs.realpath.native` - hence why it breaks on trying to find `native` since `realpath` in this case is undefined.

I am not sure how to solve that other than doing the hack described above.

**Reply by nicolo-ribaudo** [reply]
> I am trying to create a repro, but then bumped it failing on another error. However, I reported the same issue at `node-fs-extra` and got a reply (see https://github.com/jprichardson/node-fs-extra/issues/905#issuecomment-847111313) to do a `module.exports = fs` instead of `export default fs` in my fs mock. And it worked. I have no idea why.

Oh the reason is that `export default foo;` is _not_ `module.exports = foo`, but it's `exports.default = foo`.

This is because the `default` export is "just a named export with an optional shorter syntax". For example, you could do this:
```js
let a = 1, b = 2, c = 3;

export { a as default, b as x, c as y };
```

**Reply by thernstig** [reply]
So there is no solution to this, other than to use `module.exports = fs` in my mocked ECMAScript `fs` module?

I had to read up on this a bit, but is there not a way where `@babel/plugin-transform-modules-commonjs` could solve this transparently? For example https://github.com/microsoft/TypeScript/issues/2719 seems to suggest babel did an interop like:

> In order to encourage the use of CommonJS and ES6 modules, when exporting a default
export with no other exports `module.exports` will be set in addition to `exports["default"]`

```js
export default test;
```

```js
exports["default"] = test;
module.exports = exports["default"];
```

**Or alternatively**, could `@babel/plugin-transform-modules-commonjs` check if there is a native CJS module importing a transpiled ESM module, and in that case change `const fs = require('fs')` into `const fs = require('fs').default` automatically in the transpilation?

**Reply by nicolo-ribaudo** [reply]
Babel used to behave similarly to TypeScript a few years ago, but it has a very big problem: this change becomes a breaking change for people using Babel to compile their libraries:
```diff
  export default 3;
+ export const 2 = 2;
```

Anyway, you could use https://www.npmjs.com/package/babel-plugin-add-module-exports if you want that behaviour.

> **Or alternatively**, could `@babel/plugin-transform-modules-commonjs` check if there is a native CJS module importing a transpiled ESM module, and in that case change `const fs = require('fs')` into `const fs = require('fs').default` automatically in the transpilation?

Babel operates on a per-file basis, so you will need:
- a separate plugin, to which we pass a list of specifiers to convert (`["fs"]`)
- jest should be configured to pass every file (even internal Node.js files doing `require("fs")`) to Babel.

> So there is no solution to this, other than to use `module.exports = fs` in my mocked ECMAScript `fs` module?

In general, CJS and EDM have different semantics: I recommend keeping them separate, and use CJS if you want to mock a CJS module.

**Reply by thernstig** [reply]
Thanks! The unfortunate side-effect of keeping it as a CJS module means we need specific ESLint behavior for all mocks, as all other code is ESM since we use `"type": "module"` in `package.json`. Or we could use `babel-plugin-add-module-exports` as you also recommend.

There is no way to use `export { ... }` in the `fs` mock I assume without these workarounds?
